<launch>
  <!-- Eulcidean clustering Parameter -->
  <arg name="remove_ground_ransac" default="false" />
  <arg name="remove_ground_rayGroundFilter" default="true" />
  <arg name="downsample_cloud" default="false" /> <!-- Apply VoxelGrid Filter with the value given by "leaf_size"-->
  <arg name="leaf_size" default="0.2" /><!-- Voxel Grid Filter leaf size-->
  <arg name="cluster_size_min" default="20" /><!-- Minimum number of points to consider a cluster as valid-->
  <arg name="cluster_size_max" default="100000" /><!-- Maximum number of points to allow inside a cluster-->
  <arg name="sync" default="false" />
  <arg name="use_diffnormals" default="false" />
  <arg name="pose_estimation" default="true" />
  <arg name="clipCloud" default="true" />
  <arg name="clip_min_height" default="-1.3" />
  <arg name="clip_max_height" default="0.5" />
  <arg name="in_max_height" default="0.3" />
  <arg name="in_floor_max_angle" default="0.15" />

  <arg name="keep_lanes" default="true" />
  <arg name="keep_lane_left_distance" default="5" />
  <arg name="keep_lane_right_distance" default="5" />
  <arg name="cluster_merge_threshold" default="1.5" />
  <arg name="clustering_distance" default="0.75" />

  <arg name="use_vector_map_in_floor_max_angle" default="false" />
  <arg name="wayarea_gridmap_layer" default="wayarea" />

  <arg name="output_frame" default="/os1_lidar" />

  <arg name="remove_points_upto" default="2.0" />

  <arg name="use_gpu" default="false" />

  <arg name="use_multiple_thres" default="false"/>
  <arg name="clustering_ranges" default="[15,30,45,60]"/><!-- Distances to segment pointcloud -->
  <arg name="clustering_distances" default="[0.5,1.1,1.6,2.1,2.6]"/><!-- Euclidean Clustering threshold distance for each segment -->

  <!-- Ray ground Parameter --> 
  <!-- <arg name="general_max_slope" default="5.0" />    degrees -->
  <!-- <arg name="local_max_slope" default="15.0" />     degrees -->
  <!-- <arg name="radial_divider_angle" default="0.08" />    distance in rads between dividers -->
  <!-- <arg name="concentric_divider_distance" default="0.01" />    distance in meters between concentric divisions -->
  <!-- <arg name="min_height_threshold" default="0.05" />     minimum height threshold regardless the slope, useful for close points -->
  <!-- <arg name="clipping_height" default="-2.0" />    the points higher than this will be removed from the input cloud. -->
  <!-- <arg name="min_point_distance" default="1.0" />   minimum distance from the origin to consider a point as valid -->
  <!-- <arg name="reclass_distance_threshold" default="0.2" />   distance between points at which re classification will occur -->

  <arg name="general_max_slope" default="10.0" />  <!-- degrees -->
  <arg name="local_max_slope" default="15.0" />   <!-- degrees -->
  <arg name="radial_divider_angle" default="0.01" />  <!-- distance in rads between dividers -->
  <arg name="concentric_divider_distance" default="0.1" />  <!-- distance in meters between concentric divisions -->
  <arg name="min_height_threshold" default="0.1" />   <!-- minimum height threshold regardless the slope, useful for close points -->
  <arg name="clipping_height" default="0.01" />  <!-- the points higher than this will be removed from the input cloud. -->
  <arg name="min_point_distance" default="0.5" /> <!-- minimum distance from the origin to consider a point as valid -->
  <arg name="reclass_distance_threshold" default="0.2" /> <!-- distance between points at which re classification will occur -->

  <node pkg="detection" type="detection" name="euclideanClustering" output="screen">
    <!-- Eulcidean clustering Parameter -->
    <param name="remove_ground_ransac" value="$(arg remove_ground_ransac)"/>
    <param name="remove_ground_rayGroundFilter" value="$(arg remove_ground_rayGroundFilter)"/>
    <param name="downsample_cloud" value="$(arg downsample_cloud)"/>
    <param name="leaf_size" value="$(arg leaf_size)"/>
    <param name="cluster_size_min" value="$(arg cluster_size_min)"/>
    <param name="cluster_size_max" value="$(arg cluster_size_max)"/>
    <param name="use_diffnormals" value="$(arg use_diffnormals)"/>
    <param name="pose_estimation" value="$(arg pose_estimation)"/>
    <param name="keep_lanes" value="$(arg keep_lanes)"/>
    <param name="keep_lane_left_distance" value="$(arg keep_lane_left_distance)"/>
    <param name="keep_lane_right_distance" value="$(arg keep_lane_right_distance)"/>
    <param name="clipCloud" value = "$(arg clipCloud)" />
    <param name="clip_min_height" value="$(arg clip_min_height)"/>
    <param name="clip_max_height" value="$(arg clip_max_height)"/>
    <param name="output_frame" value="$(arg output_frame)"/>
    <param name="remove_points_upto" value="$(arg remove_points_upto)"/>
    <param name="clustering_distance" value="$(arg clustering_distance)"/>
    <param name="in_max_height" value="$(arg in_max_height)"/>
    <param name="in_floor_max_angle" value="$(arg in_floor_max_angle)"/>
    <param name="cluster_merge_threshold" value="$(arg cluster_merge_threshold)"/>
    <param name="use_gpu" value="$(arg use_gpu)"/>
    <param name="use_multiple_thres" value="$(arg use_multiple_thres)"/>
    <param name="clustering_ranges" value="$(arg clustering_ranges)"/><!-- Distances to segment pointcloud -->
    <param name="clustering_distances" value="$(arg clustering_distances)"/><!-- Euclidean Clustering threshold distance for each segment -->


    <!-- Ray ground Parameter -->
    <param name="general_max_slope" value="$(arg general_max_slope)"/>
    <param name="local_max_slope" value="$(arg local_max_slope)"/>
    <param name="radial_divider_angle" value="$(arg radial_divider_angle)"/>
    <param name="concentric_divider_distance" value="$(arg concentric_divider_distance)"/>
    <param name="min_height_threshold" value="$(arg min_height_threshold)"/>
    <param name="clipping_height" value="$(arg clipping_height)"/>
    <param name="min_point_distance" value="$(arg min_point_distance)"/>
    <param name="reclass_distance_threshold" value="$(arg reclass_distance_threshold)"/>


  </node>

  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find detection)/rviz/default.rviz" />

</launch>
